cpp/src/arrow/util/compression_zlib.cc

```diff
@@ -47,6 +47,238 @@ static constexpr int GZIP_CODEC = 16;
 // Determine if this is libz or gzip from header.
 static constexpr int DETECT_CODEC = 32;
 
+// ----------------------------------------------------------------------
+// gzip decompressor implementation
+
+class GZipDecompressor : public Decompressor {
+ public:
+  GZipDecompressor() : initialized_(false) {}
+
+  ~GZipDecompressor() override {
+    if (initialized_) {
+      inflateEnd(&stream_);
+    }
+  }
+
+  Status Init(GZipCodec::Format format) {
+    DCHECK(!initialized_);
+    memset(&stream_, 0, sizeof(stream_));
+    finished_ = false;
+
+    int ret;
+    // Initialize to run specified format
+    int window_bits = WINDOW_BITS;
+    if (format == GZipCodec::DEFLATE) {
+      window_bits = -window_bits;
+    } else if (format == GZipCodec::GZIP) {
+      window_bits += GZIP_CODEC;
+    }
+    if ((ret = inflateInit2(&stream_, window_bits)) != Z_OK) {
+      return ZlibError("zlib inflateInit failed: ");
+    } else {
+      initialized_ = true;
+      return Status::OK();
+    }
+  }
+
+  Status Decompress(int64_t input_len, const uint8_t* input, int64_t output_len,
+                    uint8_t* output, int64_t* bytes_read, int64_t* bytes_written,
+                    bool* need_more_output) override {
+    stream_.next_in = const_cast<Bytef*>(reinterpret_cast<const Bytef*>(input));
+    stream_.avail_in = static_cast<uInt>(input_len);
```

Comment:
When `input_len > UINT_MAX` avail_in is truncated to 0 (see https://coliru.stacked-crooked.com/a/e8e64f3d7b473acd) which will enter in a weird state. In wouldn't be surprised if users tries to decompress file greater than 4GB.

I suggest to emulate `std::clamp` with (0, std::numeric_limits<uInt>::max()).

Reply:
Right. We're generally not very careful to test for truncation :-/

Reply:
Sometimes it the chance is quite low, in this specific case (compression buffer), I think a user will hit this at least once sooner than later.